File: oop/labaratories/labaratory-work-2/Domain/Salad.cs
namespace labaratory_work_2.Domain;

public enum Dressing
{
    OliveOil,
    Mayonnaise
}

public class Salad : Dish
{
    public Dressing Dressing { get; set; } = Dressing.OliveOil;

    public Salad() : base()
    {
    }

    public Salad(string name) : base(name)
    {
    }

    public Salad(string name, int weight, double price)
        : base(name, weight, price)
    {
    }

    public void ChangeDressing()
    {
        Dressing = Dressing == Dressing.OliveOil ? Dressing.Mayonnaise : Dressing.OliveOil;
    }

    public string TossWithDressing()
    {
        return $"Tossing the salad {Name} with {Dressing}...";
    }

    public override string DisplayInfo()
    {
        return $"Salad: {Name}, {Dressing} dressing, {Weight}g, ${CalcFullPrice()}";
    }

    public override double CalcFullPrice()
    {
        return Math.Round(
           Dressing == Dressing.OliveOil ? Price * 1.4 : Price * 1.3,
           2,
           MidpointRounding.AwayFromZero
        );
    }
}

File: oop/lab-3/src/main/java/com/example/demo/domain/Salad.java
package com.example.demo.domain;

import com.example.demo.dto.DishResponse;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;
import java.math.BigDecimal;
import java.math.RoundingMode;

@Entity
@Table(name = "salads")
@PrimaryKeyJoinColumn(name = "dish_id")
@OnDelete(action = OnDeleteAction.CASCADE)
@Getter @Setter
public class Salad extends Dish {
    public enum Dressing { OLIVE_OIL, MAYONNAISE }

    @Enumerated(EnumType.STRING)
    private Dressing dressing = Dressing.OLIVE_OIL;

    public Salad() { super(); }
    public Salad(String name) { super(name); }
    public Salad(String name, int weight, double price) { super(name, weight, price); }

    public void changeDressing() {
        this.dressing = (this.dressing == Dressing.OLIVE_OIL) ? Dressing.MAYONNAISE : Dressing.OLIVE_OIL;
    }

    public String tossWithDressing() {
        return "Tossing the salad " + getName() + " with " + getDressing() + "...";
    }

    @Override
    public String displayInfo() {
        return String.format("Salad: %s, %s dressing, %dg, $%s",
                getName(), getDressing(), getWeight(), calcFullPrice());
    }

    @Override
    public double calcFullPrice() {
        double factor = (this.dressing == Dressing.OLIVE_OIL) ? 1.4 : 1.3;
        double result = getPrice() * factor;
        return BigDecimal.valueOf(result)
                .setScale(2, RoundingMode.HALF_UP)
                .doubleValue();
    }

    @Override
    public DishResponse getInfo() {
        return new DishResponse(
                getId(),
                getName(),
                getWeight(),
                getPrice(),
                "Salad",
                getDressing().name()
        );
    }
}